_posts/2020-02-26-demo-fraud-detection-2.md

```diff
@@ -0,0 +1,203 @@
+<br/>
+<i><small>Figure 1: Rule definition</small></i>
+</center>
+<br/>
+
+![](./../img/blog/patterns-blog-2/rule-dsl.png)
+The previous post covered the use of `groupingKeyNames` by `DynamicKeyFunction` to extract message keys. Parameters from the second part of this rule are used by `DynamicAlertFunction`: they define the actual logic of the performed operations and their parameters (such as the alert triggering limit). This means that the same rule must be present in both `DynamicKeyFunction` and `DynamicAlertFunction`. To achieve this result, we will use the [broadcast data distribution mechanism](https://ci.apache.org/projects/flink/flink-docs-release-1.10/dev/stream/state/broadcast_state.html) of Apache Flink. Let’s turn to the Job Graph of the system that we are building:
+<center>
+<img src="{{ site.baseurl }}/img/blog/patterns-blog-2/job-graph.png" width="800px" alt="Figure 2: Job Graph of the Fraud Detection Flink Job"/>
+<br/>
+<i><small>Figure 2: Job Graph of the Fraud Detection Flink Job</small></i>
+</center>
+<br/>
+
+![](./../img/blog/patterns-blog-2/job-graph.png)
+
+The main blocks of the Transactions processing pipeline are:<br>
+
+* **Transaction Source**, consuming transaction messages from Kafka partitions in parallel <br>
+
+* **Dynamic Key Function**, which performs data enrichment with a dynamic key. The subsequent `keyBy` operator hashes this dynamic key and partitions the data accordingly among all parallel instances of the following task
+
+* **Dynamic Alert Function** that accumulates a data window and creates Alerts based on it
+
+The job graph also depicts various data exchange patterns between the operators.
+
+* The __FORWARD__ connection following the Transaction Source means that all data consumed by one of the parallel instances of the Transaction Source operator is transferred to exactly one instance of the subsequent `DynamicKeyFunction` operator, without redistribution. It also indicates the same level of parallelism of the two connected operators (12 in this case). This communication pattern is depicted in Figure 3. Orange circles represent transactions and dotted rectangles depict parallel instances of the operators (also called [*tasks*](https://ci.apache.org/projects/flink/flink-docs-stable/concepts/glossary.html#operator)).  
+
+<center>
+<img src="{{ site.baseurl }}/img/blog/patterns-blog-2/forward.png" width="800px" alt="Figure 3: Message passing across tasks: FORWARD"/>
+<br/>
+<i><small>Figure 3: Message passing across tasks: FORWARD</small></i>
+</center>
+<br/>
+
+![](./../img/blog/patterns-blog-2/forward.png)
+
+* The __HASH__ connection between the `DynamicKeyFunction` and the `DynamicAlertFunction` means that for each message a hash code is calculated and messages are evenly distributed among available parallel instances of the following operator. Such connection needs to be explicitly "requested" from Flink by using a `keyBy` operator.
```

Comment:
```suggestion
* The __HASH__ connection between the `DynamicKeyFunction` and the `DynamicAlertFunction` means that for each message a hash code is calculated and messages are evenly distributed among available parallel instances of the following operator. Such connection needs to be explicitly "requested" from Flink by using `keyBy`.
```

Reply:
Sometimes it gets awkward with names that end with a proper noun, like "Function", because when you read them, in your head it sounds like you have to slap an article there :)